yt_dlp/plugins.py

```diff
@@ -0,0 +1,169 @@
+
+
+class PluginFinder(importlib.abc.MetaPathFinder):
+    """
+    This class provides one or multiple namespace packages
+    it searches in sys.path for the existing subdirectories
+    from which the modules can be imported
+    """
+
+    def __init__(self, *packages):
+        self.packages = set()
+        self._zip_content_cache = {}
+
+        for name in packages:
+            self.packages.update(self.partition(name))
+
+    @staticmethod
+    def partition(name):
+        yield from itertools.accumulate(name.split('.'), lambda a, b: '.'.join((a, b)))
+
+    def zip_has_dir(self, archive, path):
+        if archive not in self._zip_content_cache:
+            self._zip_content_cache[archive] = [Path(name) for name in ZipFile(archive).namelist()]
+        return any(path in file.parents for file in self._zip_content_cache[archive])
+
+    def search_locations(self, fullname):
+        # Also load plugin packages from standard config folders
+        config_locations = []
+        for config_dir in map(Path, get_user_config_dirs('yt-dlp') + get_system_config_dirs('yt-dlp')):
+            plugin_dir = config_dir / 'plugins'
+            if not plugin_dir.is_dir():
+                continue
+            config_locations.extend(plugin_dir / d for d in plugin_dir.iterdir())
+        path_locations = [Path(path) for path in sys.path]  # PYTHON_PATH
+        parts = Path(*fullname.split('.'))
+        locations = set()
+        for path in dict.fromkeys(path_locations + config_locations):
+            candidate = path / parts
+            if candidate.is_dir():
+                locations.add(str(candidate))
+            elif path.name and any(path.with_suffix(suffix).is_file() for suffix in {'.zip', '.egg', '.whl'}):
+                with contextlib.suppress(FileNotFoundError):
+                    if self.zip_has_dir(path, parts):
+                        locations.add(str(candidate))
+
+        return list(locations)
+
+    def find_spec(self, fullname, path=None, target=None):
+        if fullname not in self.packages:
+            return None
+
+        search_locations = self.search_locations(fullname)
+        if not search_locations:
+            return None
+
+        spec = importlib.machinery.ModuleSpec(fullname, PluginLoader(), is_package=True)
+        spec.submodule_search_locations = search_locations
+        return spec
+
+    def invalidate_caches(self):
+        self._zip_content_cache.clear()
+        for package in self.packages:
+            if package in sys.modules:
+                del sys.modules[package]
+
+
+def initialize():
+    global _INITIALIZED
+    if _INITIALIZED:
+        return
+
+    # FIXME: https://github.com/yt-dlp/yt-dlp/pull/1393/files#r742829806
+    # are we running from PyInstaller single executable?
+    # then copy the plugin directory if exist
+    root = Path(sys.executable).parent
+    meipass = Path(getattr(sys, '_MEIPASS', root))
+    if getattr(sys, 'frozen', False) and root != meipass:
```

Comment:
Same here, I have not tested whatever this pyinstaller integration is

Reply:
The Pyinstaller executable can only import modules from the temporary _MEIPASS path where the uncompressed files reside. That's why we need to copy the plugin files over. Maybe we postpone this feature after the other functionality is stable?

Reply:
Hmm, so this is for detecting plugins in the `ytdlp_plugins` folder that is in the same folder the exe is in?

Without it, it doesn't appear to successfully detect them. However plugins in user configuration folders still seem to work.

I think this would be required for backwards-compat, so will need to keep it in. It's at least working but I'll look into pukkandans suggestions from previous pr on it.

Reply:
Actually, I think I can get the same behaviour by adding utils.get_executable_path() to the locations it searches. That seems to allow it to work with both pyinstaller and py2exe?

Reply:
Yes, we always want "portable configuration files" to be loaded from `get_executable_path`

Additionally loading from _MEIPASS is nice, since it'll allow someone to distribute a binary with plugins included (This is what the [other comment](https://github.com/yt-dlp/yt-dlp/pull/5553?notification_referrer_id=NT_kwDOAGUQnLI0ODQzODcyNDA3OjY2MjMzODg&notifications_query=is%3Aunread#discussion_r1023331143) refers to). But we definitely don't want to copy everything to _MEIPASS

Reply:
Ah, I see. 

I think I'll leave the loading from _MEIPASS out of this PR to keep it simple - distributing a binary with the plugins included can be a future improvement imo.